src/library/scala/collection/mutable/OrderAwareHashMap.scala

```diff
@@ -0,0 +1,800 @@
+    var n = tree
+    while(n ne null) {
+      val next = n.next
+      if((n.hash & mask) == 0) { // keep low
+        lastLow.next = n
+        lastLow = n
+      } else { // move to high
+        lastHigh.next = n
+        lastHigh = n
+      }
+      n = next
+    }
+    lastLow.next = null
+    if(tree ne preLow.next) table(lowBucket) = preLow.next
+    if(preHigh.next ne null) {
+      table(highBucket) = preHigh.next
+      lastHigh.next = null
+    }
+  }
+
+  private[this] def splitBucket(tree: RBNode, lowBucket: Int, highBucket: Int, mask: Int): Unit = {
+    var lowCount, highCount = 0
+    tree.foreachNode((n: RBNode) => if((n.hash & mask) != 0) highCount += 1 else lowCount += 1)
+    if(highCount != 0) {
+      if(lowCount == 0) {
+        table(lowBucket) = null
+        table(highBucket) = tree
+      } else {
+        table(lowBucket) = fromNodes(new OrderAwareHashMap.RBNodesIterator(tree).filter(n => (n.hash & mask) == 0), lowCount)
+        table(highBucket) = fromNodes(new OrderAwareHashMap.RBNodesIterator(tree).filter(n => (n.hash & mask) != 0), highCount)
+      }
+    }
+  }
+
+  private[this] def tableSizeFor(capacity: Int) =
+    (Integer.highestOneBit((capacity-1).max(4))*2).min(1 << 30)
+
+  private[this] def newThreshold(size: Int) = (size.toDouble * loadFactor).toInt
+
+  override def clear(): Unit = {
+    java.util.Arrays.fill(table.asInstanceOf[Array[AnyRef]], null)
+    contentSize = 0
+  }
+
+  override def remove(key: K): Option[V] = remove0(key) match {
+    case null => None
+    case nd => nd
+  }
+
+  def subtractOne(elem: K): this.type = { remove0(elem); this }
+
+  override def knownSize: Int = size
+
+  override def isEmpty: Boolean = size == 0
+
+  override def foreach[U](f: ((K, V)) => U): Unit = {
+    val len = table.length
+    var i = 0
+    while(i < len) {
+      val n = table(i)
+      if(n ne null) n match {
+        case n: LLNode => n.foreach(f)
+        case n: RBNode => n.foreach(f)
+      }
+      i += 1
+    }
+  }
+
+  protected[this] def writeReplace(): AnyRef = new DefaultSerializationProxy(new OrderAwareHashMap.DeserializationFactory[K, V](table.length, loadFactor, ordering), this)
+
+  override protected[this] def stringPrefix = "OrderAwareHashMap"
+
+
+
+  /*
+  override def getOrElseUpdate(key: K, defaultValue: => V): V = {
+    val hash = computeHash(key)
+    val idx = index(hash)
+    val nd = table(idx) match {
+      case null => null
+      case nd => nd.findNode(key, hash)
+    }
+    if(nd != null) nd.value
+    else {
+      val table0 = table
+      val default = defaultValue
+      if(contentSize + 1 >= threshold) growTable(table.length * 2)
+      // Avoid recomputing index if the `defaultValue()` or new element hasn't triggered a table resize.
+      val newIdx = if (table0 eq table) idx else index(hash)
+      put0(key, default, false, hash, newIdx)
+      default
+    }
+  }
+  */
+
+  ///////////////////// RedBlackTree code derived from mutable.RedBlackTree:
```

Comment:
This is a _lot_ of duplicate code.  How big of a difference does it make to have this duplicated rather than simply reusing the hopefully-already-optimized `RedBlackTree` code?

Reply:
There are substantial differences between the two implementations. `RedBlackTree` has a separate `Tree` object at the root which gets updates when the root node changes and which also keeps track of the size. The implementation here removes this extra `Tree` object and either returns the root node from operations (in cases where they don't have to return anything else) or updates the hash table directly.

Reply:
I've taken another look to see if we could generalize the implementation here. Removing the common `Node` superclass doesn't appear to make difference. I wanted to benchmark the impact of abstracting over an interface to get and set the root node but then I realized that `RBNode` also contains the hash of the key. We'd need a lot of key + hash tuples and/or subclasses of `RBNode` to abstract over this, both of which has a high impact on performance.

Reply:
Yeah, good point, and I guess rehashing is also a bad idea.  I just worry that RB tree code is nontrivial enough, and here are in a place hard enough to test, that we might be vulnerable to bugs.  But I guess a common implementation has to wait for Valhalla so the tuples are approximately zero-cost.

Reply:
You wouldn't do any rehashing unless you convert back from tree to linked list (which we don't support anyway). The tree stores the hashes because they are used for ordering in the tree before falling back to the "real" `Ordering`. We can do this because we don't actually need the tree to be sensibly ordered and comparing hashes is usually much faster than comparing the actual keys.

Reply:
I thought the whole point was to defend against cases where all the hashes are identical anyway.  So it may be fast, but isn't it also pointless?

Reply:
A bucket can also get treeified when it contains many different hashes. Checking the hashes first helps with these bucket collisions. It doesn't do anything for the worst case (actual hash collisions).

Reply:
How can it not slow things down in the worst case by loading and checking the (identical) hash values before moving on to the real test?  In any case, it seems like it would just add complexity and if the threshold for treeifying is conservative, it would be very unusual to end up in the same-bucket-but-different-hashes case.  If we need a local impl of RBTree anyway, even if we forget hashes, then sure, may as well do the test--maybe the big win of using hashes in the few cases it matters is worth the small loss of looking at them when they're identical.  But if this is the only thing that's blocking use of the library RBTree, I think it's at least worth reconsidering dropping the hash.  (Maybe the custom impl is enough faster for that to be worth the custom copy; usually the library versions have too much inheritance baggage to actually perform at max speed.)

Reply:
I had already typed this:
> Why would you want to make the threshold for treeifying so conservative that it only happens in DoS cases? If you have more than a handful of elements in a bucket the tree will be faster than a linked list. And in these cases you benefit from checking the hash first, just like the linked list buckets benefit from keeping the list sorted by hash and also checking the hash first.

...but then I ran some more benchmarks. It turns out that the hash checks in RBNode don't buy you anything. `oaFillRegular` is the same with or without hash checks (not surprising because it uses mostly LLNodes), `oaFillDoS` is slower with hash checks (also expected because the hash checks are useless) but even `oaFillColliding` is a bit slower with hash checks. You can probably construct a test case where the version with the hash checks is faster but in most cases it looks like we're better off without.

Reply:
Ah, but we still can't remove the hashes from RBNode. They are needed for splitting the bucket!

Reply:
Oh, right, of course.  They could be recomputed, though maybe that's too expensive?  But hash maps don't always assume cached hash codes, and I'm still very tempted by the prospect of avoiding the code duplication of two RB trees.

Reply:
In the benchmarks the difference from rehashing is quite small but that's to be expected since all keys cache their hash codes anyway. For keys with expensive hash calculations this could lead to surprising performance issues because the regular LLNodes cache the hash codes and so do all of our other hash map/set implementations, and Java's as well.

Reply:
I did some preliminary refactoring that would allow abstracting over different nodes with a generic red-black-tree implementation. The performance penalty is around 10 to 20% which is quite significant.

Reply:
"A generic red-black-tree implementation" meaning the one we have in the standard library, or yet another one?  Anyway, I agree that the slowdown sounds significant (and thanks for testing!).  Then again, if it only hits the hopefully rare heavily-collided case, maybe the slowdown is acceptable if the implementation is simplified sufficiently?

Reply:
The implementation was based on the one here but allows for different node types with or without hashes and values. You would get the same performance degradation if you used it for TreeMap and TreeSet.

Reply:
I guess to my mind the only advantage comes if we _literally_ use `scala.collection.mutable.RedBlackTree` for the impacted nodes here.  Anything else involves significant code duplication; if you're already duplicating code, may as well make it ideal for the application (as you already did).